app/store/middleware.js

```diff
@@ -0,0 +1,201 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+import DeviceInfo from 'react-native-device-info';
+
+import {ViewTypes} from 'app/constants';
+import Config from 'assets/config';
+
+export function messageRetention() {
+    return (next) => (action) => {
+        if (action.type === 'persist/REHYDRATE') {
+            const {app} = action.payload;
+            const {entities} = action.payload;
+
+            if (!entities) {
+                return next(action);
+            }
+
+            // When a new version of the app has been detected
+            if (!app || !app.version || app.version !== DeviceInfo.getVersion() || app.build !== DeviceInfo.getBuildNumber()) {
+                return next(resetStateForNewVersion(action));
+            }
+
+            // Keep only the last 60 messages for the last 5 viewed channels in each team
+            // and apply data retention on those posts if applies
+            return next(cleanupState(action));
+        } else if (action.type === ViewTypes.DATA_CLEANUP) {
+            const nextAction = cleanupState(action, true);
+            return next(nextAction);
+        }
+
+        return next(action);
+    };
+}
+
+function resetStateForNewVersion(action) {
+    const {payload} = action;
+    const lastChannelForTeam = getLastChannelForTeam(payload);
+    const currentUserId = payload.entities.users.currentUserId;
+
+    const nextState = {
+        app: {
+            build: DeviceInfo.getBuildNumber(),
+            version: DeviceInfo.getVersion()
+        },
+        entities: {
+            general: payload.entities.general,
+            teams: {
+                currentTeamId: payload.entities.teams.currentTeamId,
+                teams: payload.entities.teams.teams,
+                myMembers: payload.entities.teams.myMembers
+            },
+            users: {
+                currentUserId,
+                profiles: {
+                    [currentUserId]: payload.entities.users.profiles[currentUserId]
+                }
+            },
+            preferences: payload.entities.preferences,
+            search: {
+                recent: payload.entities.search.recent
+            }
+        },
+        views: {
+            channel: {
+                drafts: payload.views.channel.drafts
+            },
+            i18n: payload.views.i18n,
+            fetchCache: payload.views.fetchCache,
+            team: {
+                lastTeamId: payload.views.team.lastTeamId,
+                lastChannelForTeam
+            },
+            thread: {
+                drafts: payload.views.thread.drafts
+            },
+            selectServer: payload.views.selectServer
+        }
+    };
+
+    return {
+        type: action.type,
+        payload: nextState,
+        error: action.error
+    };
+}
+
+function getLastChannelForTeam(payload) {
```

Comment:
Related to an earlier comment, maybe it would make sense to just change the field name so that we don't need this migration code

Reply:
see above response